include/zephyr/kernel.h

```diff
@@ -5698,13 +5698,35 @@ struct k_poll_event {
 	};
 };
 
+#define K_POLL_ASSERT_VALID_OBJECT(_event_type, _event_obj) BUILD_ASSERT_EXPR( \
+	(_event_type) == K_POLL_TYPE_IGNORE || (\
+		(_event_type) == K_POLL_TYPE_SIGNAL && \
+		__builtin_types_compatible_p(__typeof__(_event_obj), struct k_poll_signal *)\
+	) || (\
+		(_event_type) == K_POLL_TYPE_SEM_AVAILABLE && \
+		__builtin_types_compatible_p(__typeof__(_event_obj), struct k_sem *) \
+	) || (\
+		(_event_type) == K_POLL_TYPE_DATA_AVAILABLE && \
+		__builtin_types_compatible_p(__typeof__(_event_obj), struct k_fifo *) \
+	) || (\
+		(_event_type) == K_POLL_TYPE_DATA_AVAILABLE && \
+		__builtin_types_compatible_p(__typeof__(_event_obj), struct k_queue *) \
+	) || (\
+		(_event_type) == K_POLL_TYPE_MSGQ_DATA_AVAILABLE && \
+		__builtin_types_compatible_p(__typeof__(_event_obj), struct k_msgq *) \
+	) || (\
+		(_event_type) == K_POLL_TYPE_PIPE_DATA_AVAILABLE && \
+		__builtin_types_compatible_p(__typeof__(_event_obj), struct k_pipe *)\
+	), \
+	"Invalid object for k_poll")
+
 #define K_POLL_EVENT_INITIALIZER(_event_type, _event_mode, _event_obj) \
 	{ \
 	.poller = NULL, \
 	.type = _event_type, \
 	.state = K_POLL_STATE_NOT_READY, \
 	.mode = _event_mode, \
-	.unused = 0, \
+	.unused = 0 + K_POLL_ASSERT_VALID_OBJECT(_event_type, _event_obj), \
```

Comment:
GCC statement expressions[1] are the normal tool for that.  `({ STATEMENT_A; STATEMENT_B; })` executes A and then B, but then evaluates syntactically as the result of B.

[1] An extension we do allow, because it's so pervasively supported and incredibly useful.  See https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html

Reply:
Those aren't allowed outside of function scopes, which would be a weird restriction to impose.